# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGL2Context.h
# Commit: ccfa767dba64
# Full Hash: ccfa767dba644dc193e0246eb8e8ff3377e8b8a5
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-01-09 03:40:18
# Regressor Bug: 1477756
# File Overlap Count: 1
# Description:
#   Bug 1477756 - Client-side bindings mirror for precise CC, and merge similar codepaths. r=handyman
#   
#   * Context loss using RAII
#   * Move Program reflection Client-side
#   
# ==============================================================================

diff -r 71c122ac0ca7 -r ccfa767dba64 dom/canvas/WebGL2Context.h
--- a/dom/canvas/WebGL2Context.h	Wed Jan 08 22:19:14 2020 +0000
+++ b/dom/canvas/WebGL2Context.h	Wed Jan 08 22:19:16 2020 +0000
@@ -21,12 +21,12 @@
 class OwningWebGLBufferOrLongLong;
 }  // namespace dom
 
-class WebGL2Context : public WebGLContext {
+class WebGL2Context final : public WebGLContext {
+  friend class WebGLContext;
+
  public:
-  virtual ~WebGL2Context(){};
-
-  static bool IsSupported();
-  static WebGL2Context* Create() { return new WebGL2Context(); }
+  WebGL2Context(HostWebGLContext& host, const webgl::InitContextDesc& desc)
+      : WebGLContext(host, desc) {}
 
   virtual bool IsWebGL2() const override { return true; }
 
@@ -34,18 +34,10 @@
   // Buffer objects - WebGL2ContextBuffers.cpp
 
   void CopyBufferSubData(GLenum readTarget, GLenum writeTarget,
-                         WebGLintptr readOffset, WebGLintptr writeOffset,
-                         WebGLsizeiptr size);
-
- private:
-  template <typename BufferT>
-  void GetBufferSubDataT(GLenum target, WebGLintptr offset,
-                         const BufferT& data);
-
- public:
-  Maybe<UniquePtr<RawBuffer<>>> GetBufferSubData(GLenum target,
-                                                 WebGLintptr srcByteOffset,
-                                                 size_t byteLen);
+                         uint64_t readOffset, uint64_t writeOffset,
+                         uint64_t size) const;
+  void GetBufferSubData(GLenum target, uint64_t srcByteOffset,
+                        const Range<uint8_t>& dest) const;
 
   // -------------------------------------------------------------------------
   // Framebuffer objects - WebGL2ContextFramebuffers.cpp
@@ -54,33 +46,23 @@
                        GLint dstX0, GLint dstY0, GLint dstX1, GLint dstY1,
                        GLbitfield mask, GLenum filter);
 
-  virtual MaybeWebGLVariant GetFramebufferAttachmentParameter(
-      GLenum target, GLenum attachment, GLenum pname) override;
-
-  // Make the inline version from the superclass visible here.
-  using WebGLContext::GetFramebufferAttachmentParameter;
-
   void InvalidateFramebuffer(GLenum target,
-                             const nsTArray<GLenum>& attachments);
+                             const Range<const GLenum>& attachments);
   void InvalidateSubFramebuffer(GLenum target,
-                                const nsTArray<GLenum>& attachments, GLint x,
+                                const Range<const GLenum>& attachments, GLint x,
                                 GLint y, GLsizei width, GLsizei height);
   void ReadBuffer(GLenum mode);
 
   // -------------------------------------------------------------------------
   // Renderbuffer objects - WebGL2ContextRenderbuffers.cpp
 
-  Maybe<nsTArray<int32_t>> GetInternalformatParameter(GLenum target,
-                                                      GLenum internalformat,
-                                                      GLenum pname);
+  Maybe<std::vector<int32_t>> GetInternalformatParameter(GLenum target,
+                                                         GLenum internalformat,
+                                                         GLenum pname) const;
 
   // -------------------------------------------------------------------------
   // Texture objects - WebGL2ContextTextures.cpp
 
-  void TexStorage(uint8_t funcDims, GLenum target, GLsizei levels,
-                  GLenum internalFormat, GLsizei width, GLsizei height,
-                  GLsizei depth);
-
   GLint GetFragDataLocation(const WebGLProgram& prog, const nsAString& name);
 
   // GL 3.0 & ES 3.0
@@ -98,52 +80,28 @@
                              WebGLintptr offset, GLsizei instanceCount);
   */
 
-  void DrawRangeElements(GLenum mode, GLuint start, GLuint end, GLsizei count,
-                         GLenum type, WebGLintptr byteOffset) {
-    const FuncScope funcScope(*this, "drawRangeElements");
-    if (IsContextLost()) return;
-
-    if (end < start) {
-      ErrorInvalidValue("end must be >= start.");
-      return;
-    }
-
-    DrawElementsInstanced(mode, count, type, byteOffset, 1);
-  }
-
   // ------------------------------------------------------------------------
   // Multiple Render Targets - WebGL2ContextMRTs.cpp
-  /* Implemented in WebGLContext
-  void DrawBuffers(const nsTArray<GLenum>& buffers);
-  */
 
  private:
   bool ValidateClearBuffer(GLenum buffer, GLint drawBuffer,
-                           size_t availElemCount, GLuint elemOffset,
-                           GLenum funcType);
+                           webgl::AttribBaseType funcType);
 
  public:
   void ClearBufferfi(GLenum buffer, GLint drawBuffer, GLfloat depth,
                      GLint stencil);
-  void ClearBufferfv(GLenum buffer, GLint drawBuffer,
-                     const RawBuffer<const float>& src, GLuint srcElemOffset);
-  void ClearBufferiv(GLenum buffer, GLint drawBuffer,
-                     const RawBuffer<const int32_t>& src, GLuint srcElemOffset);
-  void ClearBufferuiv(GLenum buffer, GLint drawBuffer,
-                      const RawBuffer<const uint32_t>& src,
-                      GLuint srcElemOffset);
+  void ClearBufferTv(GLenum buffer, GLint drawBuffer,
+                     const webgl::TypedQuad& data);
 
   // -------------------------------------------------------------------------
   // Sampler Objects - WebGL2ContextSamplers.cpp
 
-  already_AddRefed<WebGLSampler> CreateSampler();
-  void DeleteSampler(WebGLSampler* sampler);
-  bool IsSampler(const WebGLSampler* sampler);
+  RefPtr<WebGLSampler> CreateSampler();
   void BindSampler(GLuint unit, WebGLSampler* sampler);
   void SamplerParameteri(WebGLSampler& sampler, GLenum pname, GLint param);
   void SamplerParameterf(WebGLSampler& sampler, GLenum pname, GLfloat param);
-  MaybeWebGLVariant GetSamplerParameter(const WebGLSampler& sampler,
-                                        GLenum pname);
+  Maybe<double> GetSamplerParameter(const WebGLSampler& sampler,
+                                    GLenum pname) const;
 
   // -------------------------------------------------------------------------
   // Sync objects - WebGL2ContextSync.cpp
@@ -151,30 +109,22 @@
   const GLuint64 kMaxClientWaitSyncTimeoutNS =
       1000 * 1000 * 1000;  // 1000ms in ns.
 
-  already_AddRefed<WebGLSync> FenceSync(GLenum condition, GLbitfield flags);
-  bool IsSync(const WebGLSync* sync);
-  void DeleteSync(WebGLSync* sync);
+  RefPtr<WebGLSync> FenceSync(GLenum condition, GLbitfield flags);
   GLenum ClientWaitSync(const WebGLSync& sync, GLbitfield flags,
                         GLuint64 timeout);
-  void WaitSync(const WebGLSync& sync, GLbitfield flags, GLint64 timeout);
-  MaybeWebGLVariant GetSyncParameter(const WebGLSync& sync, GLenum pname);
 
   // -------------------------------------------------------------------------
   // Transform Feedback - WebGL2ContextTransformFeedback.cpp
 
-  already_AddRefed<WebGLTransformFeedback> CreateTransformFeedback();
-  void DeleteTransformFeedback(WebGLTransformFeedback* tf);
-  bool IsTransformFeedback(const WebGLTransformFeedback* tf);
-  void BindTransformFeedback(GLenum target, WebGLTransformFeedback* tf);
+  RefPtr<WebGLTransformFeedback> CreateTransformFeedback();
+  void BindTransformFeedback(WebGLTransformFeedback* tf);
   void BeginTransformFeedback(GLenum primitiveMode);
   void EndTransformFeedback();
   void PauseTransformFeedback();
   void ResumeTransformFeedback();
   void TransformFeedbackVaryings(WebGLProgram& program,
-                                 const nsTArray<nsString>& varyings,
-                                 GLenum bufferMode);
-  Maybe<WebGLActiveInfo> GetTransformFeedbackVarying(
-      const WebGLProgram& program, GLuint index);
+                                 const std::vector<std::string>& varyings,
+                                 GLenum bufferMode) const;
 
   // -------------------------------------------------------------------------
   // Uniform Buffer Objects and Transform Feedback Buffers -
@@ -185,53 +135,22 @@
       void BindBufferRange(GLenum target, GLuint index, WebGLBuffer* buffer,
                            WebGLintptr offset, WebGLsizeiptr size);
   */
-  MaybeWebGLVariant GetParameter(GLenum pname) override;
+  Maybe<double> GetParameter(GLenum pname) override;
+  Maybe<double> GetIndexedParameter(GLenum pname, uint32_t index) const;
 
   // Make the inline version from the superclass visible here.
   using WebGLContext::GetParameter;
-  MaybeWebGLVariant GetIndexedParameter(GLenum target, GLuint index);
-  MaybeWebGLVariant GetUniformIndices(const WebGLProgram& program,
-                                      const nsTArray<nsString>& uniformNames);
-  MaybeWebGLVariant GetActiveUniforms(const WebGLProgram& program,
-                                      const nsTArray<GLuint>& uniformIndices,
-                                      GLenum pname);
 
-  GLuint GetUniformBlockIndex(const WebGLProgram& program,
-                              const nsAString& uniformBlockName);
-  MaybeWebGLVariant GetActiveUniformBlockParameter(const WebGLProgram& program,
-                                                   GLuint uniformBlockIndex,
-                                                   GLenum pname);
-  nsString GetActiveUniformBlockName(const WebGLProgram& program,
-                                     GLuint uniformBlockIndex);
   void UniformBlockBinding(WebGLProgram& program, GLuint uniformBlockIndex,
                            GLuint uniformBlockBinding);
 
-  // -------------------------------------------------------------------------
-  // Vertex Array Object - WebGL2ContextVAOs.cpp
-  // TODO(djg): Implemented in WebGLContext
-  /*
-      already_AddRefed<WebGLVertexArrayObject> CreateVertexArray();
-      void DeleteVertexArray(WebGLVertexArrayObject* vertexArray);
-      bool IsVertexArray(WebGLVertexArrayObject* vertexArray);
-      void BindVertexArray(WebGLVertexArrayObject* vertexArray);
-  */
-
  private:
-  WebGL2Context() {
-    MOZ_ASSERT(IsSupported(),
-               "not supposed to create a WebGL2Context"
-               "context when not supported");
-  }
-
   virtual UniquePtr<webgl::FormatUsageAuthority> CreateFormatUsage(
       gl::GLContext* gl) const override;
 
   virtual bool IsTexParamValid(GLenum pname) const override;
 
   void UpdateBoundQuery(GLenum target, WebGLQuery* query);
-
-  // CreateVertexArrayImpl is assumed to be infallible.
-  virtual WebGLVertexArray* CreateVertexArrayImpl() override;
 };
 
 }  // namespace mozilla